scenes/cslib_array.txt

```diff
@@ -247,6 +247,180 @@
+*temp e_idx count
+*temp store
+*label _reverse_loop
+*if (s_idx < e_idx)
+    *set store {p_array_ref&"_${s_idx}"}
+    *set {p_array_ref&"_${s_idx}"} {p_array_ref&"_${e_idx}"}
+    *set {p_array_ref&"_${e_idx}"} store
+    *set s_idx + 1
+    *set e_idx - 1
+    *goto _reverse_loop
+*return
+
+
+*comment FOR_EACH
+*comment ------------------------------
+*comment Call a given subroutine against each
+*comment element in the array.
+*comment
+*comment The given routine should expect
+*comment to be passed three parameters:
+*comment - The element value
+*comment - The element's source array index
+*comment - The name of the array
+*comment
+*comment And should return a new value for the element.
+*comment E.g.:
+*comment    *label capitalize_str
+*comment    *params value index array_ref
+*comment    *set cslib_ret "$!!{value}"
+*comment    *return
+*comment ------------------------------
+*comment    params:
+*comment        p_array_ref(string): the name of the array to iterate over
+*comment        p_scene(string): the name of the scene containing the routine
+*comment        p_label(string): the name of the label denoting the start of the routine
+*label for_each
+*params p_array_ref p_scene p_label
+*set cslib_ret ""
+*temp count {p_array_ref&"_count"}
+*temp n 1
+*label _for_each_loop
+*if (n <= count)
+    *gosub_scene {p_scene} {p_label} {p_array_ref&"_${n}"} n p_array_ref
+    *set {p_array_ref&"_${n}"} cslib_ret
+    *set n + 1
+    *goto _for_each_loop
+*return
+
+
+*comment TEST_ALL
+*comment ------------------------------
+*comment Apply a given condition subroutine against each
+*comment element in the array and return a boolean
+*comment denoting whether or not ALL elements met it.
+*comment
+*comment The given routine should expect
+*comment to be passed three parameters:
+*comment - The element value
+*comment - The element's source array index
+*comment - The name of the array
+*comment
+*comment And should return a boolean (true = condition met).
+*comment E.g.:
+*comment    *label is_short_string
+*comment    *params value index
+*comment    *set cslib_ret false
+*comment    *if (length(value) < 10)
+*comment        *set cslib_ret true
+*comment    *return
+*comment ------------------------------
+*comment    params:
+*comment        p_array_ref(string): the name of the array to iterate over
+*comment        p_scene(string): the name of the scene containing the routine
+*comment        p_label(string): the name of the label denoting the start of the routine
+*comment    returns:
+*comment        (boolean): true if all elements match the condition (or array is empty), else false
+*label test_all
+*params p_array_ref p_scene p_label
+*temp count {p_array_ref&"_count"}
+*temp n 1
+*label _test_all_loop
+*if (n <= count)
+    *gosub_scene {p_scene} {p_label} {p_array_ref&"_${n}"} n p_array_ref
+    *if not(cslib_ret)
+        *return
+    *set n + 1
+    *goto _test_all_loop
+*set cslib_ret true
+*return
+
+
+*comment TEST_SOME
```

Comment:
Maybe "test\_any"?

Reply:
I can go with that!
On the topic: test or check?

Reply:
🤔

Reply:
test 😂